plugins/com.google.cloud.tools.eclipse.appengine.localserver.test/src/com/google/cloud/tools/eclipse/appengine/localserver/server/LocalAppEngineServerBehaviourTest.java

```diff
@@ -0,0 +1,153 @@
+      serverBehavior.checkAndSetPorts(portProber);
+      fail();
+    } catch (CoreException ex) {
+      assertEquals("Port must be between 0 and 65535.", ex.getMessage());
+    }
+  }
+
+  @Test
+  public void testCheckAndSetPorts_outOfBoundServerPort() {
+    try {
+      serverBehavior.serverPort = 65536;
+      serverBehavior.checkAndSetPorts(portProber);
+      fail();
+    } catch (CoreException ex) {
+      assertEquals("Port must be between 0 and 65535.", ex.getMessage());
+    }
+  }
+
+  @Test
+  public void testCheckAndSetPorts_serverPortInUse() {
+    try {
+      when(portProber.isPortInUse(8080)).thenReturn(true);
+
+      serverBehavior.serverPort = 8080;
+      serverBehavior.checkAndSetPorts(portProber);
+      fail();
+    } catch (CoreException ex) {
+      assertEquals("Port 8080 is in use.", ex.getMessage());
+    }
+  }
+
+  @Test
+  public void testCheckAndSetPorts_adminPortInUse() throws CoreException {
+    when(portProber.isPortInUse(8000)).thenReturn(true);
+
+    serverBehavior.serverPort = 65535;
+    serverBehavior.checkAndSetPorts(portProber);
+    assertEquals(65535, serverBehavior.serverPort);
+    assertEquals(0, serverBehavior.adminPort);
+  }
+
+  @Test
+  public void testExtractPortFromUrl_hostName() {
+    int port = LocalAppEngineServerBehaviour.extractPortFromUrl("http://localhost:5678");
+    assertEquals(5678, port);
+
+    port = LocalAppEngineServerBehaviour.extractPortFromUrl("http://my-machine:1234");
+    assertEquals(1234, port);
+
+    port = LocalAppEngineServerBehaviour.extractPortFromUrl("http://www.server-example.com:80");
+    assertEquals(80, port);
+  }
+
+  @Test
+  public void testExtractPortFromUrl_ipv4Address() {
+    int port = LocalAppEngineServerBehaviour.extractPortFromUrl("http://0.0.0.0:5678");
+    assertEquals(5678, port);
+
+    port = LocalAppEngineServerBehaviour.extractPortFromUrl("http://192.168.1.4:1234");
+    assertEquals(1234, port);
+  }
+
+  @Test
+  public void testExtractPortFromUrl_noPortUrl() {
+    int port = LocalAppEngineServerBehaviour.extractPortFromUrl("http://localhost");
+    assertEquals(-1, port);
+  }
+
+  @Test
+  public void testExtractPortFromUrl_noMatch() {
+    int port = LocalAppEngineServerBehaviour.extractPortFromUrl("arbitrary string");
+    assertEquals(-1, port);
+  }
+
+  private static final String[] serverOutput = new String[] {
+      "WARNING  2016-11-03 21:11:21,930 devappserver2.py:785] DEFAULT_VERSION_HOSTNAME will not be set correctly with --port=0",
+      "INFO     2016-11-03 21:11:21,956 api_server.py:205] Starting API server at: http://localhost:52892",
+      "INFO     2016-11-03 21:11:21,959 dispatcher.py:197] Starting module \"default\" running at: http://localhost:55948",
+      "INFO     2016-11-03 21:11:21,959 admin_server.py:116] Starting admin server at: http://localhost:43679",
+      "Nov 03, 2016 9:11:23 PM com.google.appengine.tools.development.SystemPropertiesManager setSystemProperties"
```

Comment:
Might want to update with the multiple-service console text in one of my other comments so as to verify that you're pulling out the right default service port.